src/neurostatslib/solver.py

```diff
@@ -0,0 +1,593 @@
+"""
+## A Module for Optimization with Various Regularizations.
+
+This module provides a series of classes that facilitate the optimization of models
+with different types of regularizations. Each solver class in this module interfaces
+with various optimization methods, and they can be applied depending on the model's requirements.
+"""
+import abc
+import inspect
+from typing import Callable, List, Optional, Tuple, Union
+
+import jax.numpy as jnp
+import jaxopt
+from numpy.typing import NDArray
+
+from .base_class import _Base
+from .proximal_operator import prox_group_lasso
+
+__all__ = ["UnRegularizedSolver", "RidgeSolver", "LassoSolver", "GroupLassoSolver"]
+
+
+def __dir__() -> list[str]:
+    return __all__
+
+
+class Solver(_Base, abc.ABC):
+    """
+    Abstract base class for optimization solvers.
+
+    This class is designed to provide a consistent interface for optimization solvers,
+    enabling users to easily switch between different solvers and ensure compatibility
+    with various loss functions and regularization schemes.
+
+    Attributes
+    ----------
+    allowed_optimizers :
+        List of optimizer names that are allowed for use with this solver.
```

Comment:
"solver" and "optimizers" are a bit ambiguous to me, do we have / can we write a description describing the difference between the two? By optimizers, these are different optimization algorithms, right? whereas solvers largely implement regularization?

Reply:
solver is a more general class, which include solvers of minimization problems (optimizers). 
We are only talking about optimizers, so our classes are actually just objects calling appropriately the optimizers.
We can re-label `allowed_optimizers` to `allowed_algorithms`.
The alternative is to call the module differently, since it is a wrapper to specific regularization, so maybe "regularizer.py"?  and rename all the classes to regularization. 

what solution is better?

What the class does for now is:

- create a solver function which combines a loss and a penalization

which is in between a just applying a regularizer and defining a solver. 

Both operation needs to be performed:
1. We need to wrap jaxopt because the API is not the same for proximal method vs smooth methods so that GLM can seamlessly call the same thing. 
2. We need to define appropriately the penalization.

The regularization strategy and the penalization itself are related. it would be nice if we were able to split the two, to keep it more unix like, any idea on how to do that?

Reply:
So renaming `Solver` -> `Regularizer`, and `allowed_optimizers` -> `allowed_solvers` (to match jaxopt's [terminology](https://jaxopt.github.io/stable/unconstrained.html))

Reply:
I renamed everything, I tried to update all the docstrings and notes with the new names, but I would double check that I spotted everything